Mathlib/Data/Matroid/IndepAxioms.lean

```diff
@@ -0,0 +1,565 @@
+
+    -- We have `|I₀ + e| ≤ |J|`, since otherwise we could extend the maximal set `J`
+    have hcard : (insert e I₀).ncard ≤ J.ncard
+    · refine not_lt.1 fun hlt ↦ ?_
+      obtain ⟨f, hfI, hfJ, hfi⟩ := ind_aug hJ hJfin heI₀i (hI₀fin.insert e) hlt
+      have hfE₀ : f ∈ E₀ := mem_of_mem_of_subset hfI (insert_subset_insert (subset_union_left _ _))
+      refine hfJ (insert_eq_self.1 <| Eq.symm (hJmax _
+        ⟨hB₀J.trans <| subset_insert _ _,hfi,insert_subset hfE₀ hJss⟩ (subset_insert _ _)))
+
+    -- But this means `|I₀| < |J|`, and extending `I₀` into `J` gives a contradiction
+    rw [ncard_insert_of_not_mem heI₀ hI₀fin, ←Nat.lt_iff_add_one_le] at hcard
+
+    obtain ⟨f, hfJ, hfI₀, hfi⟩ := ind_aug (ind_mono hI hI₀I) hI₀fin hJ hJfin hcard
+    exact hI₀ f ⟨Or.elim (hJss hfJ) (fun hfe ↦ (heJ <| hfe ▸ hfJ).elim) (by aesop), hfI₀⟩ hfi )
+
+    ( by
+      rintro X - I hI hIX
+      have hzorn := zorn_subset_nonempty {Y | Indep Y ∧ I ⊆ Y ∧ Y ⊆ X} ?_ I ⟨hI, Subset.rfl, hIX⟩
+      · obtain ⟨J, hJ, -, hJmax⟩ := hzorn
+        exact ⟨J, hJ, fun K hK hJK ↦ (hJmax K hK hJK).subset⟩
+
+      refine fun Is hIs hchain ⟨K, hK⟩ ↦ ⟨⋃₀ Is, ⟨?_,?_,?_⟩, fun _ ↦ subset_sUnion_of_mem ⟩
+      · refine h_compact _ fun J hJ hJfin ↦ ?_
+        have hchoose : ∀ e, e ∈ J → ∃ I, I ∈ Is ∧ (e : α) ∈ I
+        · exact fun _ he ↦ mem_sUnion.1 <| hJ he
+        choose! f hf using hchoose
+        refine J.eq_empty_or_nonempty.elim (fun hJ ↦ hJ ▸ h_empty) (fun hne ↦ ?_)
+        obtain ⟨x, hxJ, hxmax⟩ := Finite.exists_maximal_wrt f _ hJfin hne
+        refine ind_mono (hIs (hf x hxJ).1).1 fun y hyJ ↦ ?_
+        obtain (hle | hle) := hchain.total (hf _ hxJ).1 (hf _ hyJ).1
+        · rw [hxmax _ hyJ hle]; exact (hf _ hyJ).2
+        exact hle (hf _ hyJ).2
+
+      · exact subset_sUnion_of_subset _ K (hIs hK).2.1 hK
+      exact sUnion_subset fun X hX ↦ (hIs hX).2.2 )
+    h_support
+
+@[simp] theorem matroidOfIndepOfFinitary_apply (E : Set α) (Indep : Set α → Prop)
+    (h_empty : Indep ∅)
+    (ind_mono : ∀ ⦃I J⦄, Indep J → I ⊆ J → Indep I)
+    (ind_aug : ∀ ⦃I J⦄, Indep I → I.Finite → Indep J → J.Finite → I.ncard < J.ncard →
+      ∃ e ∈ J, e ∉ I ∧ Indep (insert e I))
+    (h_compact : ∀ I, (∀ J, J ⊆ I → J.Finite → Indep J) → Indep I)
+    (h_support : ∀ ⦃I⦄, Indep I → I ⊆ E) :
+    (matroidOfIndepOfFinitary E Indep h_empty ind_mono ind_aug h_compact h_support).Indep
+      = Indep := by
+  simp [matroidOfIndepOfFinitary]
+
+@[simp] theorem matroidOfIndepOfFinitary_ground (E : Set α) (Indep : Set α → Prop)
+    (h_empty : Indep ∅)
+    (ind_mono : ∀ ⦃I J⦄, Indep J → I ⊆ J → Indep I)
+    (ind_aug : ∀ ⦃I J⦄, Indep I → I.Finite → Indep J → J.Finite → I.ncard < J.ncard →
+      ∃ e ∈ J, e ∉ I ∧ Indep (insert e I))
+    (h_compact : ∀ I, (∀ J, J ⊆ I → J.Finite → Indep J) → Indep I)
+    (h_support : ∀ ⦃I⦄, Indep I → I ⊆ E) :
+    (matroidOfIndepOfFinitary E Indep h_empty ind_mono ind_aug h_compact h_support).E = E := by
+  simp [matroidOfIndepOfFinitary]
```

Comment:
isn't this true by `rfl`?

Reply:
no, because `(Matroid.ofIndepOfFinitary E Indep ... _).Indep` isn't defeq to `Indep`.

Reply:
This is in turn true because `Matroid` is defined in terms of `Base` predicate rather than an `Indep` predicate. The latter would have given better defeq properties, given that independence is the usual way to construct a matroid. Changing this would need a (manageable) refactor, but that's for another day.